### PR TITLE
WIP: Validation

### PR DESCRIPTION
**Pull request by jml**
Depends on #2

---

This change is [Reviewable](https://reviewable.io/reviews/jml/graphql-api/3)

### CODE REVIEW

File: src/GraphQL/Output.hs [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+--   * be a map
+--   * have a "data" key iff the operation executed
+--   * have an "errors" key iff the operation encountered errors
+--   * not include "data" if operation failed before execution (e.g. syntax errors,
+--     validation errors, missing info)
+--   * not have keys other than "data", "errors", and "extensions"
+--
+-- Other interesting things:
+--
+--   * Doesn't have to be JSON, but does have to have maps, strings, lists,
```
Comment (teh) [inline src/GraphQL/Output.hs]: interestingly it's almost like a protobuf - and I'm sure we could encode it as that (but for internal RPC sth like a `Binary` encoding might be faster)

File: src/GraphQL/Output.hs [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+--
+--   * must be a non-empty list
+--   * each error is a map with "message", optionally "locations" key
+--     with list of locations
+--   * locations are maps with 1-indexed "line" and "column" keys.
+data Response
+  = Success Map
+  | PreExecutionFailure Errors
+  | ExecutionFailure Errors
+  | PartialSuccess Map Errors
```
Comment (teh) [inline src/GraphQL/Output.hs]: `PartialSuccess` doesn't make sense if errors/data is either/or (i.e. an `Either`) as described above

File: src/GraphQL/Validation.hs [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+union HumanOrAlien = Human | Alien
+
+type QueryRoot {
+  dog: Dog
+}
+-}
+
+newtype ValidDocument = Valid AST.Document deriving (Eq, Show)
+
+validate :: Alternative m => AST.Document -> m ValidDocument
```
Comment (teh) [inline src/GraphQL/Validation.hs]: what's the purpose of validate? Make sure the query is valid? Matches the types?

File: src/GraphQL/Value.hs [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+fieldSetToMap (FieldSet fields) = Map (Map.fromList [ (name, value) | Field name value <- toList fields ])
+
+empty :: FieldSet
+empty = FieldSet Set.empty
+
+singleton :: Field -> FieldSet
+singleton = FieldSet . Set.singleton
+
+-- TODO: Fail on duplicate keys.
+union :: Alternative m => FieldSet -> FieldSet -> m FieldSet
```
Comment (teh) [inline src/GraphQL/Value.hs]: is `union` related to graphql union types? If so I'm not sure why it should fail on duplicates

File: src/GraphQL/Output.hs (hunk 1 above)
Comment (jml) [inline src/GraphQL/Output.hs]: Yep. Or thrift, which is I think what FB use.

File: src/GraphQL/Output.hs (hunk 2 above)
Comment (jml) [inline src/GraphQL/Output.hs]: I'm pretty sure it does make sense (although maybe my comment doesn't). 

> If the data entry in the response is not null, the errors entry in the response may contain any errors that occurred during execution. If errors occurred during execution, it should contain those errors.

From https://facebook.github.io/graphql/#sec-Errors

Note that `data` can be absent or null, and there are different meanings associated with each.

**Comment by jml** [comment]
I'll wait until I've had a chance to review your work and talk to you about it before I merge this.

**Comment by teh** [comment]
Feel free to merge - I think this is good as is

**Pull request merged**